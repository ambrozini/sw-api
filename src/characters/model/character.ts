export enum Episodes {
  NEWHOPE = "NEWHOPE",
  EMPIRE = "EMPIRE",
  JEDI = "JEDI",
}

type Planet = string;

export interface Character {
  name: string;
  episodes: Episodes[];
  planet?: Planet;
}
