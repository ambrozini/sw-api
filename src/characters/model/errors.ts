export const CHARACTER_SERVICE_ERRORS = [
  "ALREADY_EXISTS",
  "VALIDATION_ERROR",
  "NOT_EXISTS",
] as const;

export type CharacterServiceErrors = typeof CHARACTER_SERVICE_ERRORS;
