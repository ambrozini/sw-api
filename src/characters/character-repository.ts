import { drop, take } from "lodash";
import { Character, Episodes } from "./model/character";

export class CharacterRepository {
  private static instance: CharacterRepository;

  private characters: Character[] = [];

  private constructor() {
    this.clearData();
  }

  static getInstance(): CharacterRepository {
    if (!this.instance) {
      this.instance = new CharacterRepository();
    }

    return this.instance;
  }

  clearData() {
    this.characters = [
      {
        name: "Luke Skywalker",
        episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
      },
      {
        name: "Darth Vader",
        episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
      },
      {
        name: "Han Solo",
        episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
      },
      {
        name: "Leia Organa",
        episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
        planet: "Alderaan",
      },
      {
        name: "Wilhuff Tarkin",
        episodes: [Episodes.NEWHOPE],
      },
      {
        name: "C-3PO",
        episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
      },
      {
        name: "R2-D2",
        episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
      },
    ];
  } // TODO remove when switching into real db

  find({
    limit = 5,
    offset = 0,
  }: {
    limit: number;
    offset: number;
  }): Character[] {
    return take(drop([...this.characters], offset), limit);
  }

  findOne(userName: string): Character | undefined {
    return this.characters.find((character) => character.name === userName);
  }

  exists(predicate: (character: Character) => boolean): boolean {
    return this.characters.some((character) => predicate(character));
  }

  create(character: Character) {
    this.characters.push(character);
  }

  delete(name: string) {
    this.characters = this.characters.filter(
      (character) => character.name !== name
    );
  }

  update(characterToUpdate: Character) {
    this.characters = [
      ...this.characters.filter(
        (character) => character.name !== characterToUpdate.name
      ),
      characterToUpdate,
    ];
  }
}
