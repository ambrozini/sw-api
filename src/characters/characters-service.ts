import { ActionResult, errorAction, successAction } from "@shared";
import { Character, Episodes } from "./model/character";
import { CharacterServiceErrors } from "./model/errors";

const characters: Character[] = [
  {
    name: "Luke Skywalker",
    episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
  },
  {
    name: "Darth Vader",
    episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
  },
  {
    name: "Han Solo",
    episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
  },
  {
    name: "Leia Organa",
    episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
    planet: "Alderaan",
  },
  {
    name: "Wilhuff Tarkin",
    episodes: [Episodes.NEWHOPE],
  },
  {
    name: "C-3PO",
    episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
  },
  {
    name: "R2-D2",
    episodes: [Episodes.NEWHOPE, Episodes.EMPIRE, Episodes.JEDI],
  },
];

export const getAll = (): Character[] => {
  return [...characters];
};

export const create = (
  character: Character
): ActionResult<null, CharacterServiceErrors> => {
  const characterList = getAll();

  if (
    characterList.some(
      (existingCharacter) => existingCharacter.name === character.name
    )
  ) {
    return errorAction("ALREADY_EXISTS");
  }

  if (!character.name) {
    return errorAction("VALIDATION_ERROR", "Name doesn't exists");
  }

  if (!character.episodes) {
    return errorAction("VALIDATION_ERROR", "Episodes list doesn't exists");
  }

  if (character.episodes.length === 0) {
    return errorAction(
      "VALIDATION_ERROR",
      "Character have to be at least in one episode"
    );
  }

  characterList.push(character);

  return successAction();
};
