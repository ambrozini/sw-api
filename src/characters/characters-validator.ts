import { errorAction, Validator } from "@shared";
import { Character, Episodes } from "./model/character";
import { CharacterServiceErrors } from "./model/errors";

export const commonValidator: Validator<Character, CharacterServiceErrors> = (
  character: Character
) => {
  if (!character.name) {
    return errorAction("VALIDATION_ERROR", "Name doesn't exists");
  }

  if (!character.episodes) {
    return errorAction("VALIDATION_ERROR", "Episodes list doesn't exists");
  }

  if (character.episodes.length === 0) {
    return errorAction(
      "VALIDATION_ERROR",
      "Character have to be at least in one episode"
    );
  }

  if (
    character.episodes.some(
      (episode) => !Object.values(Episodes).includes(episode)
    )
  ) {
    return errorAction("VALIDATION_ERROR", "Such episode doesn't exist");
  }

  return null;
};
