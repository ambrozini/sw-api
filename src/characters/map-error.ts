import { ErrorResponse, ErrorResult, textHeader } from "@shared";
import { StatusCodes } from "http-status-codes";
import { CharacterServiceErrors } from "./model/errors";

export function mapError(
  error: ErrorResult<CharacterServiceErrors>
): ErrorResponse {
  const headers = textHeader;

  switch (error?.errorCode) {
    case "ALREADY_EXISTS":
      return {
        statusCode: StatusCodes.CONFLICT,
        body: "Character with such name already exists",
        headers,
      };
    case "VALIDATION_ERROR":
      return {
        statusCode: StatusCodes.BAD_REQUEST,
        body: error.message,
        headers,
      };

    case "NOT_EXISTS":
      return {
        statusCode: StatusCodes.NOT_FOUND,
        body: "Such character doesn't exists",
        headers,
      };

    default:
      throw new Error("Unkown Error");
  }
}
