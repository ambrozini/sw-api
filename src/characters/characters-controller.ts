import { Response, isActionResultFailure, Error } from "@shared";
import { StatusCodes } from "http-status-codes";

import * as service from "./characters-service";
import { Character } from "./model/character";
import { CharacterServiceErrors } from "./model/errors";

export const getAll = async (): Promise<Response> => {
  const characters = service.getAll();

  return {
    statusCode: 200,
    body: JSON.stringify({ characters }),
  };
};

export const create = async ({
  body: character,
}: {
  body: Character;
}): Promise<Response> => {
  const result = service.create(character);

  if (isActionResultFailure(result)) {
    return mapError(result);
  }

  return {
    statusCode: 201,
  };
};

function mapError(error: Error<CharacterServiceErrors>): Response {
  switch (error.errorCode) {
    case "ALREADY_EXISTS":
      return {
        statusCode: StatusCodes.CONFLICT,
        error: "Character with such name already exists",
      };
    case "VALIDATION_ERROR":
      return {
        statusCode: StatusCodes.BAD_REQUEST,
        error: error.message,
      };
  }
}
