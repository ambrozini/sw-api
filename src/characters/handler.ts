import { Response, middify } from "@shared";

import * as controller from "./characters-controller";
import { Character } from "./model/character";

export const find = middify<null, Response<Character[]>>(controller.find);

export const findOne = middify<null, Response<Character>>(controller.findOne);

export const create = middify<Character, Response>(controller.create);

export const deleteOne = middify<{ name: string }, Response>(
  controller.deleteOne
);

export const update = middify<Character, Response>(controller.update);
