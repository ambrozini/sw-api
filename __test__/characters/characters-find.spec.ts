import { HttpEvent, SuccessResponse } from "@shared";
import { find } from "../../src/characters/handler";
import { noop } from "lodash";
import { Character } from "src/characters/model/character";

describe("Characters Integration Tests - find", () => {
  describe("with default limitation", () => {
    let response: SuccessResponse<Character[]>;

    beforeEach(async () => {
      response = (await find(
        {
          httpMethod: "GET",
        } as HttpEvent<null>,
        null,
        noop
      )) as SuccessResponse<Character[]>;
    });
    it("should return successfull response", () => {
      expect(response.statusCode).toEqual(200);
    });

    it("should return 5 entries", () => {
      expect(response.body.length).toEqual(5);
    });

    it("should return first 5 entries", () => {
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("with limit 2", () => {
    let response: SuccessResponse<Character[]>;

    beforeEach(async () => {
      response = (await find(
        {
          httpMethod: "GET",
          queryStringParameters: {
            limit: "2",
          },
        } as unknown as HttpEvent<null>,
        null,
        noop
      )) as SuccessResponse<Character[]>;
    });
    it("should return successfull response", () => {
      expect(response.statusCode).toEqual(200);
    });

    it("should return 2 entries", () => {
      expect(response.body.length).toEqual(2);
    });
  });

  describe("with no number limit", () => {
    let response: SuccessResponse<Character[]>;

    beforeEach(async () => {
      response = (await find(
        {
          httpMethod: "GET",
          queryStringParameters: {
            limit: "no_number",
          },
        } as unknown as HttpEvent<null>,
        null,
        noop
      )) as SuccessResponse<Character[]>;
    });
    it("should return 400 status code", () => {
      expect(response.statusCode).toEqual(400);
    });

    it("should return error message", () => {
      expect(response.body).toEqual("Limit has to be integer");
    });
  });

  describe("with offset of 2", () => {
    let response: SuccessResponse<Character[]>;

    beforeEach(async () => {
      response = (await find(
        {
          httpMethod: "GET",
          queryStringParameters: {
            offset: "2",
          },
        } as unknown as HttpEvent<null>,
        null,
        noop
      )) as SuccessResponse<Character[]>;
    });
    it("should return successfull response", () => {
      expect(response.statusCode).toEqual(200);
    });

    it("should return 5 entries", () => {
      expect(response.body.length).toEqual(5);
    });

    it("should return entries between 2 and 7", () => {
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("with no number offset", () => {
    let response: SuccessResponse<Character[]>;

    beforeEach(async () => {
      response = (await find(
        {
          httpMethod: "GET",
          queryStringParameters: {
            offset: "no_number",
          },
        } as unknown as HttpEvent<null>,
        null,
        noop
      )) as SuccessResponse<Character[]>;
    });
    it("should return 400 error code", () => {
      expect(response.statusCode).toEqual(400);
    });

    it("should return error message", () => {
      expect(response.body).toEqual("Offset has to be integer");
    });
  });

  describe("read two pages", () => {
    let firstPageResponse: SuccessResponse<Character[]>;
    let secondPageResponse: SuccessResponse<Character[]>;

    beforeEach(async () => {
      firstPageResponse = (await find(
        {
          httpMethod: "GET",
        } as unknown as HttpEvent<null>,
        null,
        noop
      )) as SuccessResponse<Character[]>;

      secondPageResponse = (await find(
        {
          httpMethod: "GET",
          queryStringParameters: {
            offset: firstPageResponse.body.length,
          },
        } as unknown as HttpEvent<null>,
        null,
        noop
      )) as SuccessResponse<Character[]>;
    });

    it("should return first page", () => {
      expect(firstPageResponse.body).toMatchSnapshot();
    });

    it("should return second page", () => {
      expect(secondPageResponse.body).toMatchSnapshot();
    });
  });
});
