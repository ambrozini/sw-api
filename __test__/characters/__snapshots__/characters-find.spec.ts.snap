// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters Integration Tests - find read two pages should return first page 1`] = `
[
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Luke Skywalker",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Darth Vader",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Han Solo",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Leia Organa",
    "planet": "Alderaan",
  },
  {
    "episodes": [
      "NEWHOPE",
    ],
    "name": "Wilhuff Tarkin",
  },
]
`;

exports[`Characters Integration Tests - find read two pages should return second page 1`] = `
[
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "C-3PO",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "R2-D2",
  },
]
`;

exports[`Characters Integration Tests - find with default limitation should return first 5 entries 1`] = `
[
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Luke Skywalker",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Darth Vader",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Han Solo",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Leia Organa",
    "planet": "Alderaan",
  },
  {
    "episodes": [
      "NEWHOPE",
    ],
    "name": "Wilhuff Tarkin",
  },
]
`;

exports[`Characters Integration Tests - find with offset of 2 should return entries between 2 and 7 1`] = `
[
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Han Solo",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "Leia Organa",
    "planet": "Alderaan",
  },
  {
    "episodes": [
      "NEWHOPE",
    ],
    "name": "Wilhuff Tarkin",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "C-3PO",
  },
  {
    "episodes": [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "name": "R2-D2",
  },
]
`;
